File: lib/core/db/tag_dao.dart
import 'package:sqflite/sqflite.dart';
import '../../shared/models/tag.dart';
import '../../shared/models/media_item.dart';
import 'database.dart';

class TagDao {
  Future<Database> get _db => AppDatabase.instance;

  Future<List<Tag>> findBySpace(MediaSpace space) async {
    final db = await _db;
    final rows = await db.query(
      'tags',
      where: 'space = ?',
      whereArgs: [space.name],
      orderBy: 'is_custom ASC, label ASC',
    );
    return rows.map(Tag.fromMap).toList();
  }

  Future<int> insert(Tag tag) async {
    final db = await _db;
    return db.insert(
      'tags',
      tag.toMap(),
      conflictAlgorithm: ConflictAlgorithm.ignore,
    );
  }

  Future<List<Tag>> findByMediaId(int mediaId) async {
    final db = await _db;
    final rows = await db.rawQuery(
      '''
      SELECT t.* FROM tags t
      JOIN media_tags mt ON t.id = mt.tag_id
      WHERE mt.media_id = ?
      ORDER BY t.label ASC
    ''',
      [mediaId],
    );
    return rows.map(Tag.fromMap).toList();
  }

  Future<List<Tag>> findAll() async {
    final db = await _db;
    final rows = await db.query(
      'tags',
      orderBy: 'space ASC, is_custom ASC, label ASC',
    );
    return rows.map(Tag.fromMap).toList();
  }

  Future<int> delete(int id) async {
    final db = await _db;
    // media_tags는 ON DELETE CASCADE로 자동 삭제
    return db.delete('tags', where: 'id = ?', whereArgs: [id]);
  }

  Future<int> update(Tag tag) async {
    final db = await _db;
    return db.update('tags', tag.toMap(), where: 'id = ?', whereArgs: [tag.id]);
  }

  Future<void> setMediaTags(int mediaId, List<int> tagIds) async {
    final db = await _db;
    await db.transaction((txn) async {
      await txn.delete(
        'media_tags',
        where: 'media_id = ?',
        whereArgs: [mediaId],
      );
      for (final tagId in tagIds) {
        await txn.insert('media_tags', {
          'media_id': mediaId,
          'tag_id': tagId,
        }, conflictAlgorithm: ConflictAlgorithm.ignore);
      }
    });
  }
}

File: android-native/app/src/main/java/com/mebonsoft/memorix/core/database/dao/TagDao.kt
package com.mebonsoft.memorix.core.database.dao

import androidx.room.Dao
import androidx.room.Insert
import androidx.room.OnConflictStrategy
import androidx.room.Query
import androidx.room.Transaction
import com.mebonsoft.memorix.core.database.entity.MediaTagCrossRef
import com.mebonsoft.memorix.core.database.entity.MediaSpace
import com.mebonsoft.memorix.core.database.entity.TagEntity
import kotlinx.coroutines.flow.Flow

data class TagUsageSummary(
    val id: Long,
    val key: String,
    val label: String,
    val colorHex: String,
    val iconName: String,
    val usageCount: Int,
)

data class MediaTagAssignment(
    val mediaId: Long,
    val tagId: Long,
    val label: String,
)

data class ManagedTagSummary(
    val id: Long,
    val label: String,
    val usageCount: Int,
)

@Dao
interface TagDao {
    @Query("SELECT * FROM tags ORDER BY label ASC")
    fun observeTags(): Flow<List<TagEntity>>

    @Query(
        """
        SELECT
            tags.id AS id,
            tags.label AS label,
            COUNT(DISTINCT CASE
                WHEN media_items.batchGroupId != '' THEN media_items.batchGroupId
                ELSE 'legacy-' || media_items.id
            END) AS usageCount
        FROM tags
        LEFT JOIN media_tags ON tags.id = media_tags.tagId
        LEFT JOIN media_items ON media_items.id = media_tags.mediaId
        GROUP BY tags.id, tags.label
        ORDER BY tags.label ASC
        """,
    )
    fun observeManagedTags(): Flow<List<ManagedTagSummary>>

    @Query(
        """
        SELECT tags.* FROM tags
        INNER JOIN media_tags ON tags.id = media_tags.tagId
        WHERE media_tags.mediaId = :mediaId
        ORDER BY tags.label ASC
        """,
    )
    fun observeTagsForMedia(mediaId: Long): Flow<List<TagEntity>>

    @Query(
        """
        SELECT media_tags.mediaId AS mediaId, tags.id AS tagId, tags.label AS label
        FROM media_tags
        INNER JOIN tags ON tags.id = media_tags.tagId
        ORDER BY tags.label ASC
        """,
    )
    fun observeMediaTagAssignments(): Flow<List<MediaTagAssignment>>

    @Query(
        """
        SELECT
            tags.id AS id,
            tags.`key` AS `key`,
            tags.label AS label,
            tags.colorHex AS colorHex,
            tags.iconName AS iconName,
            COUNT(DISTINCT CASE
                WHEN media_items.batchGroupId != '' THEN media_items.batchGroupId
                ELSE 'legacy-' || media_items.id
            END) AS usageCount
        FROM tags
        INNER JOIN media_tags ON tags.id = media_tags.tagId
        INNER JOIN media_items ON media_items.id = media_tags.mediaId
        WHERE media_items.isTrashed = 0 AND media_items.isSecret = 0
        GROUP BY tags.id, tags.`key`, tags.label, tags.colorHex, tags.iconName
        ORDER BY usageCount DESC, tags.label ASC
        LIMIT :limit
        """,
    )
    fun observeTopTags(limit: Int = 10): Flow<List<TagUsageSummary>>

    @Query(
        """
        SELECT
            tags.id AS id,
            tags.`key` AS `key`,
            tags.label AS label,
            tags.colorHex AS colorHex,
            tags.iconName AS iconName,
            COUNT(DISTINCT CASE
                WHEN media_items.batchGroupId != '' THEN media_items.batchGroupId
                ELSE 'legacy-' || media_items.id
            END) AS usageCount
        FROM tags
        INNER JOIN media_tags ON tags.id = media_tags.tagId
        INNER JOIN media_items ON media_items.id = media_tags.mediaId
        WHERE media_items.isTrashed = 0
          AND media_items.isSecret = 0
          AND media_items.space = :space
        GROUP BY tags.id, tags.`key`, tags.label, tags.colorHex, tags.iconName
        ORDER BY usageCount DESC, tags.label ASC
        LIMIT :limit
        """,
    )
    fun observeTopTagsForSpace(space: MediaSpace, limit: Int = 10): Flow<List<TagUsageSummary>>

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun insert(tag: TagEntity): Long

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun replaceMediaTags(crossRefs: List<MediaTagCrossRef>)

    @Transaction
    suspend fun setMediaTags(mediaId: Long, tagIds: List<Long>) {
        clearMediaTags(mediaId)
        if (tagIds.isNotEmpty()) {
            replaceMediaTags(tagIds.distinct().map { tagId -> MediaTagCrossRef(mediaId = mediaId, tagId = tagId) })
        }
    }

    @Query("DELETE FROM media_tags WHERE mediaId = :mediaId")
    suspend fun clearMediaTags(mediaId: Long)

    @Query("DELETE FROM media_tags WHERE tagId = :tagId")
    suspend fun clearTagAssignments(tagId: Long)

    @Query("DELETE FROM tags WHERE id = :tagId")
    suspend fun deleteTagById(tagId: Long)

    @Transaction
    suspend fun deleteManagedTag(tagId: Long) {
        clearTagAssignments(tagId)
        deleteTagById(tagId)
    }
}
